# RNA-seq analysis of *P. astreoides* reciprocal transplantation experiment

The analysis of the transcriptional response following the *in situ* reciprocal transplantation of *P. astreoides* colonies between shallow and mesophotic reefs.

### Samples
To analyze the transcriptional response of *P. astreoides*, we sequenced 31 samples belonging to different time points and depths. In summary, we had 6 replicates of deep (D) and 6 replicates of shallow (S) colonies before translocation (T0), 5 replicates of translocated colonies: deep &rarr; deep (DD), deep &rarr; shallow (DS) and shallow &rarr; shallow (SS), and 4 replicates of shallow &rarr; deep (SD) translocated colonies. Paired 150-bp reads were obtained using Illumina platform. 

### Computational resources, tools and scripts
All computational-intensive tasks were performed on [Hive2](https://hivehpc.haifa.ac.il/) computer cluster of the Faculty of Natural Sciences at University of Haifa. Necessary tools were installed locally through miniconda using [this](https://docs.vultr.com/how-to-install-miniconda-on-ubuntu-22-04) guide. For each task, a slurm script was created and executed via `sbatch`. 

### [Quality control and filtering](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/QC_Filtering) 
The initial QC analysis of raw data was performed using FastQC v0.12.1: [fastqc.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/QC_Filtering/fastqc.sh). MultiQC v1.0 was used to create a report: [multiqc.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/QC_Filtering/multiqc.sh). After inspecting the results, raw reads were trimmed with Trimmomatic v0.39 using a custom adaptor [file](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/QC_Filtering/Sequencing_adaptors.fasta): [trimmomatic.multi.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/QC_Filtering/trimmomatic.multi.sh). The resulting read quality was accessed again with FastQC and [summarized](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/QC_Filtering/results) with MultiQC. The performance of rRNA depletion was tested with SortMeRNA v.4.3.6 tool with [sortmerna.multi.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/QC_Filtering/sortmerna.multi.sh) and [summarized](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/QC_Filtering/results) with MultiQC. 

### [Mapping](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/Mapping)
Filtered reads were aligned to *P. astreoides* genome [SRR19144705](https://www.ncbi.nlm.nih.gov/sra/?term=SRR19144705) using STAR v2.7.10b mapper. Firstly, genome index was created using genome fasta file and gff3 [annotation](https://osf.io/f7sc5) file. For that, gff3 file was converted to gtf using gffread v0.12.7: [gffread.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/gffread.sh). After that, the genome was indexed: [index.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/index.sh) and reads were aligned to the genome with  `--quantMode GeneCounts` option allowing to count the number reads per gene while mapping: [mapping.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/mapping.sh). Alignment [statistics](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/results/multiqc.star_report.html) were summarized using MultiQC. The presence of 3' coverage bias was accessed with RSeQC v.2.6.4 `geneBody_coverage.py` function ([rseqc.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/rseqc.sh)), and no 3' skew was [detected](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/Mapping/results).

### [Zooxantellae identification](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/Zoox_identification)
Reads, that did not align to *P. astreoides* genome, were extracted to fasta on the previous step with `--outReadsUnmapped Fastx` argument in [mapping.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/mapping.sh) script. After that, [DIAMOND aligner](https://github.com/bbuchfink/diamond) v.0.9.19.120 was used to map unaligned reads to the proteome database of available symbionts species. Firstly, the database was created and indexed with [diamonddb.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/diamonddb.sh). Then, concatenated reads were aligned to the database with [diamond.run.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/diamond.run.sh) script and `--evalue 0.01` parameter. `.m8` output files were analysed, and the counts of reads of each symbiont species was calculated for each sample by R script [zoox_clade_analysis.part1.R](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/zoox_clade_analysis.part1.R) running on server via [clade.analysis.R.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/clade.analysis.R.sh). Then, [frequency tables and heatmaps](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/Zoox_identification/results) were generated with [zoox_clade_analysis.part2.github.R](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/zoox_clade_analysis.part2.github.R). Details on the database and the scripts can be found in [readme](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/README.md). Additional exploration of unmapped reads was performed using Kraken v2.0.8 with core nt database: [kraken2.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/kraken2.sh). KronaTools v2.8.1 was used in [krona.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Zoox_identification/krona.sh) to generate hierarchical taxonomic plots for each sample.

### [Differential expression analysis](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/DE_analysis)
Raw counts tables produced with STAR in [mapping.sh](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/Mapping/mapping.sh) script were combined and analyzed with DESeq2 v.1.44 in R v.4.4.2 with [DESeq2.Pastreoides.R](https://github.com/talimass/Cayman-translocation/blob/main/RNAseq_analysis/DE_analysis/DESeq2.Pastreoides.R) script. The effect of depth was analyzed with an account to the origin of the samples, since some translocated corals originated from the same T0 coral. The analysis [produced](https://github.com/talimass/Cayman-translocation/tree/main/RNAseq_analysis/DE_analysis/results) PCA plots and distance heatmaps, as well as lists of significantly DE genes from each contrast.   
